File: rust-port/src/unilogin.rs
use scraper::{Html, Selector};
use reqwest;
use std::collections::HashMap;
use reqwest::{Response, Client};

async fn post_form(prev_r: Response, data: &HashMap<&str, &str>, client: &Client) -> Result<Response, reqwest::Error> {
    let body = prev_r.text().await?;
    let document = Html::parse_document(&body);
    let selector = Selector::parse("form").unwrap();
    let form = document.select(&selector).next().unwrap();
    let action = form.value().attr("action").unwrap();

    println!("action: {}", action);

    let res: Response = client.post(action).json(&data).send().await?;
    Ok(res)
}

pub async fn unilogin(username: &str, password: &str) -> Result<Client, reqwest::Error> {
    let client = reqwest::Client::builder()
        .user_agent("Mozilla/5.0")
        .cookie_store(true)
        .build()?;
    

    let resp = client.get("https://www.aula.dk/auth/login.php?type=unilogin")
        .send()
        .await?;
    println!("{}", resp.status());
    print!("{:?}", resp.cookies().next());
    let mut r: Response = post_form(resp, &HashMap::from([("selectedIdp", "uni_idp")]), &client).await?;
    // r = post_form(r, &HashMap::from([("username", username), ("password", password)]), &client).await?;

    let body = r.text().await?;
    println!("{}", body);

    Ok(client)
}

File: rust-port/unilogin.py
import httpx
import json
from bs4 import BeautifulSoup


def postForm(prevR, data, session):
    soup = BeautifulSoup(prevR.text, 'html.parser')
    action = soup.find("form").get("action")

    r = session.post(action, data=data, follow_redirects=True)
    return r

def unilogin(username, password):
    session = httpx.Client()
    session.headers['user-agent'] = 'Mozilla/5.0'
    resp = session.get("https://www.aula.dk/auth/login.php?type=unilogin", follow_redirects=False)

    resp = session.get(resp.headers["location"], follow_redirects=False)
    
    href = resp.headers["location"]
    r = session.get(href, follow_redirects=False)

    r = postForm(r, {"selectedIdp": "uni_idp"}, session)

    r = postForm(r, {"username": username}, session)
    r = postForm(r, {"password": password, "username": ""}, session)
    html = BeautifulSoup(r.text, 'html.parser')

    payload = {html.find("input", {"name": "SAMLResponse"}).get("name"): html.find("input", {"name": "SAMLResponse"}).get("value"),
                    html.find("input", {"name": "RelayState"}).get("name"): html.find("input", {"name": "RelayState"}).get("value")}
    # https://broker.unilogin.dk/auth/realms/broker/broker/uni_idp/endpoint

    r = session.post("https://broker.unilogin.dk/auth/realms/broker/broker/uni_idp/endpoint", data=payload, follow_redirects=True)

    r = postForm(r, {}, session)

    soup = BeautifulSoup(r.text, 'html.parser')
    action = soup.find("form", {"name": "saml-post-binding"}).get("action")
    
    html = BeautifulSoup(r.text, 'html.parser')
    payload = {html.find("input", {"name": "SAMLResponse"}).get("name"): html.find("input", {"name": "SAMLResponse"}).get("value"),
                    html.find("input", {"name": "RelayState"}).get("name"): html.find("input", {"name": "RelayState"}).get("value")}
    # https://broker.unilogin.dk/auth/realms/broker/broker/uni_idp/endpoint

    r = session.post(action, data=payload, follow_redirects=True)
    
    return session